### PR TITLE
Helm 3 Tests

### LINKED ISSUE #290: Tests fail with helm 3
Our bats tests use `helm template -x` to render only the template under test, however that flag is removed in helm 3 and it doesn't look like there are immediate plans to bring it back (https://github.com/helm/helm/issues/6290)

### PR DESCRIPTION
**Pull request by lkysow**
Fixes #290 

Requires changing how we test empty templates due to https://github.com/helm/helm/issues/7295 (`-x` is replaced with `-s` which now errors out if the rendered template is empty). Run tests in both helm 2 and 3.

### CODE REVIEW

File: templates/client-config-configmap.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
+{{- if (or (and (ne (.Values.client.enabled | toString) "-") .Values.client.enabled) (and (eq (.Values.client.enabled | toString) "-") .Values.global.enabled)) }}
```
Comment (ishustava) [inline templates/client-config-configmap.yaml:1]: Is this change intentional?

File: test/unit/_helpers.bash [hunk 2]
```diff
@@ -2,3 +2,39 @@
...
+helm() {
+  if [[ $(v2) ]]; then
+    command helm template -x "${@:3}"
+  else
+    command helm "$@"
+  fi
+}
+
+# Usage: assert_empty helm template -s <template> [flags] .
+# assert_empty makes is possible to test that a template is not rendered in
```
Comment (ishustava) [inline test/unit/_helpers.bash:19]:
```suggestion
# assert_empty makes it possible to test that a template is not rendered in
```

**Review by ishustava — DISMISSED** [review]
Looks good to me! Left a couple of minor comments.

File: templates/client-config-configmap.yaml (hunk 1 above)
Comment (lkysow) [reply templates/client-config-configmap.yaml:1]: Yes, we need to move all comments inside the `if` statement. Otherwise the template gets rendered (with just the comments) and we can't do the same helm 3 error assertion. Since some templates have comments and some don't, it's better to standardize on all templates having comments inside the enclosing `if` statements so they always result in the same helm 3 error.

Comment (ishustava) [reply templates/client-config-configmap.yaml:1]: Cool, makes sense. Thanks for clarifying!

File: test/unit/_helpers.bash [hunk 3]
```diff
@@ -2,3 +2,41 @@
 chart_dir() {
     echo ${BATS_TEST_DIRNAME}/../..
 }
+
+# helm is used to intercept the helm command in tests and change flags depending
```
Comment (ishustava) [inline test/unit/_helpers.bash:6]:
```suggestion
# helm is used to intercept the Helm command in tests and change flags depending
```

File: test/unit/client-daemonset.bats [hunk 4]
```diff
@@ -401,28 +397,26 @@ load _helpers
...
   cd `chart_dir`
   local actual=$(helm template \
-      -x templates/client-daemonset.yaml  \
+      -s templates/client-daemonset.yaml  \
       . | tee /dev/stderr |
       yq '.spec.template.spec.containers[] | select(.name=="consul") | .env[] | select(.name == "GOSSIP_KEY") | length > 0' | tee /dev/stderr)
   [ "${actual}" = "" ]
 }
 
 @test "client/DaemonSet: gossip encryption disabled in client DaemonSet when clients are disabled" {
```
Comment (ishustava) [inline test/unit/client-daemonset.bats:406]: This test is confusing to me. It seems a bit redundant given that every feature is disabled when clients are disabled, and I don't think we test this for every feature. I'd delete it, but up to you!

File: test/unit/client-daemonset.bats [hunk 5]
```diff
@@ -14,7 +14,7 @@ load _helpers
 @test "client/DaemonSet: enable with global.enabled false" {
```
Comment (ishustava) [inline test/unit/client-daemonset.bats:14]:
```suggestion
@test "client/DaemonSet: enabled with global.enabled=false and client.enabled=true" {
```

File: test/unit/client-daemonset.bats [hunk 6]
```diff
@@ -24,28 +24,24 @@ load _helpers
 
 @test "client/DaemonSet: disable with client.enabled" {
```
Comment (ishustava) [inline test/unit/client-daemonset.bats:25]:
```suggestion
@test "client/DaemonSet: disabled with client.enabled=false" {
```

File: test/unit/client-daemonset.bats [hunk 7]
```diff
@@ -24,28 +24,24 @@ load _helpers
...
+  assert_empty helm template \
+      -s templates/client-daemonset.yaml  \
       --set 'client.enabled=false' \
-      . | tee /dev/stderr |
-      yq 'length > 0' | tee /dev/stderr)
-  [ "${actual}" = "false" ]
+      .
 }
 
 @test "client/DaemonSet: disable with global.enabled" {
```
Comment (ishustava) [inline test/unit/client-daemonset.bats:33]:
```suggestion
@test "client/DaemonSet: disabled with global.enabled=false and client.enabled='-'" {
```

File: test/unit/client-snapshot-agent-serviceaccount.bats [hunk 8]
```diff
@@ -4,17 +4,13 @@ load _helpers
 
 @test "client/SnapshotAgentServiceAccount: disabled by default" {
   cd `chart_dir`
-  local actual=$(helm template \
-      -x templates/client-snapshot-agent-serviceaccount.yaml  \
-      . | tee /dev/stderr |
-      yq 'length > 0' | tee /dev/stderr)
-  [ "${actual}" = "false" ]
+  assert_empty helm template -s templates/client-snapshot-agent-serviceaccount.yaml .
```
Comment (ishustava) [inline test/unit/client-snapshot-agent-serviceaccount.bats:7]: wow! kind of cool that it became a one-liner 😄

File: test/unit/server-acl-init-job.bats [hunk 9]
```diff
@@ -185,20 +179,21 @@ load _helpers
...
 # aclBindingRuleSelector/global.acls.manageSystemACLs
 
-@test "serverACLInit/Job: no acl-binding-rule-selector flag by default" {
+@test "serverACLInit/Job: acl-binding-rule-selector flag set by default" {
   cd `chart_dir`
   local actual=$(helm template \
-      -x templates/server-acl-init-job.yaml \
+      -s templates/server-acl-init-job.yaml \
+      --set 'global.acls.manageSystemACLs=true' \
       --set 'connectInject.aclBindingRuleSlector=foo' \
```
Comment (ishustava) [inline test/unit/server-acl-init-job.bats:187]: Hmm, this test implies that we shouldn't be setting aclBindingRuleSelector. It's strange that it's passing 🤔

Comment (ishustava) [reply test/unit/server-acl-init-job.bats:187]: Oh there's a typo in `aclBindingRuleSlector`, that's why it works!

**Review by ishustava — APPROVED** [review]
Looks great 💃 💃 💃 I left a couple of minor comments, but won't block the approval.

File: test/unit/_helpers.bash (hunk 3 above)
Comment (lkysow) [reply test/unit/_helpers.bash:6]: I'm gonna use
```
`helm`
```

here to be clear that we're intercepting the literal helm string.

File: test/unit/client-daemonset.bats (hunk 4 above)
Comment (lkysow) [reply test/unit/client-daemonset.bats:406]: agreed!

File: test/unit/server-acl-init-job.bats (hunk 9 above)
Comment (lkysow) [reply test/unit/server-acl-init-job.bats:187]: nice catch, I deleted the `--set 'connectInject.aclBindingRuleSlector=foo' \`

File: test/unit/client-snapshot-agent-serviceaccount.bats (hunk 8 above)
Comment (lkysow) [reply test/unit/client-snapshot-agent-serviceaccount.bats:7]: haha yes!
